riscv/triggers.cc

```diff
@@ -282,6 +282,61 @@ void mcontrol6_t::tdata1_write(processor_t * const proc, const reg_t val, const
     timing = 0;
 }
 
+std::optional<match_result_t> icount_t::detect_icount_match(processor_t * const proc) noexcept
+{
+  if (!common_match(proc))
+    return std::nullopt;
+
+  std::optional<match_result_t> ret = std::nullopt;
+  if (pending) {
+    pending = 0;
+    hit = true;
+    ret = match_result_t(TIMING_BEFORE, action);
+  }
+
+  if (count >= 1) {
+    if (count == 1)
+      pending = 1;
+    count = count - 1;
+  }
+
+  return ret;
+}
+
+reg_t icount_t::tdata1_read(const processor_t * const proc) const noexcept
+{
+  auto xlen = proc->get_xlen();
+  reg_t tdata1 = 0;
+  tdata1 = set_field(tdata1, CSR_ICOUNT_TYPE(xlen), CSR_TDATA1_TYPE_ICOUNT);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_DMODE(xlen), dmode);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_VS, proc->extension_enabled('H') ? vs : 0);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_VU, proc->extension_enabled('H') ? vu : 0);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_HIT, hit);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_COUNT, count);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_M, m);
+  tdata1 = set_field(tdata1, CSR_ICOUNT_PENDING, pending);
```

Comment:
I think `csrr` will be a problem here, because it is supposed to read the pre-decremented value. From the Zicsr spec:

> Some CSRs, such as the instructions-retired counter, instret, may be modified as side effects of instruction execution. In these cases, if a CSR access instruction reads a CSR, it reads the value prior to the execution of the instruction.

It seems to me that `count` and `pending` have already been updated by the time this function is called, so it will read the updated value.

Reply:
You can see in `wide_counter_csr_t::unlogged_write()` how we handle this for `minstret` and friends. I think you'll need to do something like that, though it will be more complicated.

Reply:
The ```wide_counter_csr_t``` assumes each instruction will bump the counter unconditionally, which is correct for the ```minstret``` and ```mcycle```. Unfortunately, the ```icount.count``` countdowns conditionally. For instance, when writing n to ```minstret```, the ```wide_counter_csr_t``` updates the CSR to (n-1), and the next immediate read will get n after bumping. However, when writing n to ```icount.count```, if we update the CSR to (n+1), the next read should be n but can be (n+1) if it does not match. Thus, the technique of ```minstret``` does not work for ```icount.count```. (This issue exists on the Sscofpmf's mhpmcounters with conditional bumping.)

How about adding new variables to track the reading value? https://github.com/riscv-software-src/riscv-isa-sim/pull/1220/commits/57a954417e53a2b50fa86a73d7f03dd0695b0635

Reply:
Yes, that's what I meant by "more complicated". :)

I think your solution will work but I don't (yet) have a good way to test it. I'm working on that but it might take a while, so I won't hold up this PR for that.

Of course a277f69 should be squashed back.

Reply:
OK, if something goes wrong later, I will provide another PR.